hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/Path.java

```diff
@@ -214,6 +213,12 @@ public Path(String pathString) throws IllegalArgumentException {
     // uri path is the rest of the string -- query & fragment not supported
     String path = pathString.substring(start, pathString.length());
 
+    // add "./" in front of Linux relative paths so that a path containing
+    // a colon e.q. "a:b" will not be interpreted as scheme "a".
+    if (!WINDOWS && !path.isEmpty() && path.charAt(0) != '/') {
```

Comment:
Will it be better to change the condition to `if (!WINDOWS && scheme == null && colon != -1)`? I also have a concern regarding `WINDOWS`, as ":" is file system dependent, not OS specific, but this concern is beyond this PR and JIRA.

Reply:
What if it is: `/abc:bcd`. scheme == null, colon != -1? I don't think we want to add `/` once more. It will be than: `.//abc:bcd`

Reply:
`(scheme == null && slash > colon)`? What should happen in case of "./abc" and "./abc:"?

Reply:
sorry, I don't understand what you are implying. slash can be after colon - there could be many slashes and many colons.

Reply:
I refer to the first slash and the first colon in the passed `pathString` parameter:
```
int colon = pathString.indexOf(':');
int slash = pathString.indexOf('/');
```
My understanding is that prepending "./" to the path, ensures that "/" is before ":" to avoid interpreting character sequence that `Path` does not recognized as scheme to be interpreted as scheme by `URI`. If there is no colon or slash is already before a colon, there is no need to add "./".

Reply:
Yes we have more info to make the check.  What is the advantage of this change?
Three params ctor does have this check in place as well.

Reply:
To avoid String concatenation and creation of unnecessary String objects. Prior to the patch, there was no "./" in case of a long path and it can (should?) be avoided with the patch if that long path does not contain ":".
I am not a committer for Hadoop and it is a recommendation to consider only.